n2y/plugins/footnotes.py

```diff
@@ -0,0 +1,106 @@
+import re
+import logging
+
+from pandoc.types import Note, Str, Para
+
+from n2y.rich_text import TextRichText
+from n2y.blocks import ParagraphBlock
+from n2y.errors import UseNextClass
+
+
+plugin_data_key = "n2y.plugins.footnotes"
+
+logger = logging.getLogger(__name__)
+
+
+class ParagraphWithFootnoteBlock(ParagraphBlock):
+    def __init__(self, client, notion_data, page, get_children=True):
+        super().__init__(client, notion_data, page, get_children)
+        if self._is_footnote():
+            self._attach_footnote_data()
+        else:
+            raise UseNextClass()
+
+    def to_pandoc(self):
+        return None
+
+    def _attach_footnote_data(self):
+        if plugin_data_key not in self.page.plugin_data:
+            self.page.plugin_data[plugin_data_key] = {}
+        if self._footnote() not in self.page.plugin_data[plugin_data_key]:
+            self.page.plugin_data[plugin_data_key][self._footnote()] = self._footnote_ast()
+            if self._footnote_ast_empty():
+                logger.warning('Empty footnote "[%s]"', self._footnote())
+        else:
+            msg = 'Multiple footnotes for "[%s]". Skipping latest (%s)'
+            logger.warning(msg, self._footnote(), self.notion_url)
+
+    def _is_footnote(self):
+        return self._footnote() is not None
+
+    def _footnote(self):
+        first_str = self.rich_text.to_plain_text().split(" ")[0]
+        footnotes = re.findall(r"\[(\d+)\]:", first_str)
+        if len(footnotes) != 1:
+            return None
+        return footnotes[0]
+
+    def _footnote_ast(self):
+        ast = super().to_pandoc()
+        if isinstance(ast, list):
+            first_paragraph_footnote_stripped = Para(ast[0][0][2:])
+            remaining_paragraphs = ast[1:]
+            return [first_paragraph_footnote_stripped] + remaining_paragraphs
+        else:
+            paragraph_footnote_stripped = Para(ast[0][2:])
+            return paragraph_footnote_stripped
+
+    def _footnote_ast_empty(self):
+        ast = self._footnote_ast()
+        return len(ast[0][0]) == 0 if isinstance(ast, list) else len(ast[0]) == 0
+
+
+class TextRichTextWithFootnoteRef(TextRichText):
+    def __init__(self, client, notion_data, block=None):
+        super().__init__(client, notion_data, block)
+        if not self._is_footnote():
+            raise UseNextClass()
+
+    def to_pandoc(self):
+        pandoc_ast = []
+        for token in super().to_pandoc():
+            ref = self._footnote_from_token(token)
+            if ref is None:
+                pandoc_ast.append(token)
+                continue
+            if ref not in self.block.page.plugin_data[plugin_data_key]:
+                pandoc_ast.append(token)
+                logger.warning('Missing footnote "[%s]". Rendering as plain text', ref)
```

Comment:
I'd add a notion url to the block that this occurs in, so someone can easily locate where the issue occurred